### PR TITLE
Server: prevent denial of service from malformed packet

### PR DESCRIPTION
**Pull request by schroeder-**
See #1013. Prevent packets with invalid content from tricking the internal buffer and trigger a denial of service.

### CODE REVIEW

File: tests/test_connections.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # coding: utf-8
 import asyncio
 import pytest
-
-from asyncua import Client, Server
+import asyncio
+from asyncua import Client, Server, ua
 from asyncua.ua.uaerrors import BadMaxConnectionsReached
-
+import struct
```
Comment (oroulet) [inline tests/test_connections.py:7]: import struct must be in first import group at the top

File: asyncua/server/binary_server_asyncio.py [hunk 2]
```diff
@@ -65,15 +65,22 @@ def data_received(self, data):
                 try:
                     header = header_from_binary(buf)
                 except NotEnoughData:
-                    logger.debug('Not enough data while parsing header from client, waiting for more')
+                    # a packet should at least contain a header otherwise it is malformed (8 or 12 bytes)
+                    logger.debug('Not enough data while parsing header from client, empty the buffer')
+                    self._buffer = b''
```
Comment (oroulet) [inline asyncua/server/binary_server_asyncio.py:70]: that class handle one connection so one client, if we receive data that do not even have a header, aren't we fucked and can as well close connection?

File: asyncua/server/binary_server_asyncio.py [hunk 3]
```diff
@@ -65,15 +65,22 @@ def data_received(self, data):
...
+                    logger.error(f'Got malformed header {header}')
+                    self._buffer = b''
+                else:
+                    if len(buf) < header.body_size:
+                        logger.debug('We did not receive enough data from client. Need %s got %s', header.body_size,
+                                    len(buf))
+                        return
+                    # we have a complete message
+                    self.messages.put_nowait((header, buf))
+                    self._buffer = self._buffer[(header.header_size + header.body_size):]
```
Comment (oroulet) [inline asyncua/server/binary_server_asyncio.py:83]: thinking loud here but maybe we should just close connection in all these cases

**Comment by oroulet** [comment]
@schroeder- rebase?

File: asyncua/server/binary_server_asyncio.py (hunk 2 above)
Comment (schroeder-) [reply asyncua/server/binary_server_asyncio.py]: I think closing the connection, so the client can retry is a good idea.

**Comment by oroulet** [comment]
looks good but need rebase

**Comment by schroeder-** [comment]
Done

**Comment by kohtala** [comment]
Hi. That was not a rebase, but merge. Looking at the nine commits, it would be best to squash the commits into one and make the commit based on the head of master branch. Just force push the one commit on your fix branch.

Please, we use this and we scan for vulnerabilities. It gives an alarm. A fix would be welcome. A clean pull request could speed up getting the fix in a release.

**Comment by oroulet** [comment]
made a rebase here https://github.com/FreeOpcUa/opcua-asyncio/pull/1039

**Comment by kohtala** [comment]
Thank you for this. Is there information when we could expect a release with the fix?

**Pull request closed without merging**